share/wake/lib/core/json.wake

```diff
@@ -538,3 +560,359 @@ export def lastValueInList (values: List JValue): Result JValue Error =
             | flatten
             | deduplicateJObjectKeys lastValueInList
             | rmap JObject
+
+# If the input is a string, returns a singelton list
+# containing that string. This can be used to extract
+# a string from a JValue or in list operation chains
+# as a safe JValue to String conversion.
+#
+# Examples:
+# ```
+#   jString (JString "foo") = "foo", Nil
+#   jString (JInteger 10) = Nil
+#   single json |> jField "name" |> jString = "Alice", Nil
+#   single json |> jField "age" |> jString = Nil
+# ```
+export def jString: JValue => List String = match _
+    JString s = single s
+    _         = Nil
+
+# TODO: Fix this doc
+# If `input` is a JString, updates the string using `editFn`.
+#
+# Parameters:
+#   - `editFn`: The function to update a string with
+#   - `input`: the JValue to be modified
+#
+# Examples:
+# ```
+#   editJString ("{_}man") (JString "bat") = JString "batman"
+#   editJString ("{_}man") (JInteger 10) = JInteger 10
+#   (editJField "name" $ editJString ("I'm {_}")) batman = JObject ("name" :-> JString "I'm batman", Nil)
+#   def rename (newName: String): (person: JValue) => JValue = editJField "name" $ editJString (\_ newName)
+# ```
+export def editJString (editFn: String => Result String Error): (input: JValue) => Result JValue Error = match _
+    JString s =
+        require Pass es = editFn s
+        Pass (JString es)
+    _         = failWithError "attempt to use editJString on non-JString"
+
+# If the input is an integer, returns a singelton list
+# containing that integer. This can be used to extract
+# an integer from a JValue or in list operation chains
+# as a safe JValue to Integer conversion. This works
+# even if the underlying value is represented as a
+# Double as long as the double has no fractional part.
+#
+# Examples:
+# ```
+#   jInteger (JInteger 10) = 10, Nil
+#   jInteger (JDouble 10.0) = 10, Nil
+#   jInteger (JString "foo") = Nil
+#   jInteger (JDouble 3.14) = Nil
+#   single json |> jField "age" |> jInteger = 28, Nil
+#   single json |> jField "name" |> jInteger = Nil
+# ```
+export def jInteger: JValue => List Integer = match _
+    JInteger i = single i
+    JDouble d =
+        def Pair whole frac = dmodf d
+        if (dabs frac) ==. 0.0 then
+            whole, Nil
```

Comment:
Unneeded `()`s and didn't use `single`.

Reply:
For that matter, `dabs` is not needed:
```
-.0.0 ==. 0.0: Boolean = True
```

Reply:
(but nice job considering signed zero -- I had to double-check if IEEE says those should be equal or not)

Reply:
Ah perfect! I didn't realize floating point equality worked in edge cases, I thought the C++ compiler was free to make that into a integer equality statement. TIL I made a bad assumption.